src/Microsoft.ML.Recommend/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,403 @@
+                if (!type.Equals(_parent.InputXType))
+                {
+                    throw Contracts.ExceptParam(nameof(schema), "Input X type '{0}' incompatible with predictor X type '{1}'",
+                        type, _parent.InputXType);
+                }
+                type = schema.GetColumnType(yCol);
+                if (!type.Equals(_parent.InputYType))
+                {
+                    throw Contracts.ExceptParam(nameof(schema), "Input Y type '{0}' incompatible with predictor Y type '{1}'",
+                        type, _parent.InputYType);
+                }
+            }
+
+            private Delegate[] CreateGetter(IRow input, bool[] active)
+            {
+                Contracts.CheckValue(input, nameof(input));
+                Contracts.Assert(Utils.Size(active) == OutputSchema.ColumnCount);
+
+                var getters = new Delegate[1];
+                if (active[0])
+                {
+                    CheckInputSchema(input.Schema, _xColIndex, _yColIndex);
+                    var xGetter = input.GetGetter<uint>(_xColIndex);
+                    var yGetter = input.GetGetter<uint>(_yColIndex);
+                    getters[0] = _parent.GetGetter(xGetter, yGetter);
+                }
+                return getters;
+            }
+
+            public IRow GetRow(IRow input, Func<int, bool> predicate, out Action disposer)
+            {
+                var active = Utils.BuildArray(OutputSchema.ColumnCount, predicate);
+                var getters = CreateGetter(input, active);
+                disposer = null;
+                return new SimpleRow(OutputSchema, input, getters);
+            }
+
+            public ISchemaBindableMapper Bindable { get { return _parent; } }
+        }
+
+        public int UserFeaturesSize { get { return -1; } }
+        public KeyType UserIdType { get { return _inputYType; } }
+        public KeyType ItemIdType { get { return _inputXType; } }
+
+        public UserHistoryToItemsMapper GetRecommendMapper(int recommendationCount, bool includeHistory)
+        {
+            _host.CheckParam(recommendationCount > 0, nameof(recommendationCount), "Recommendation count must be positive");
+
+            var scores = new List<KeyValuePair<uint, Float>>(_n);
+            var history = new HashSet<uint>();
+
+            UserHistoryToItemsMapper mapper =
+                (ref uint userId, ref VBuffer<float> userFeatures, ref VBuffer<uint> items, ref VBuffer<float> weights,
+                    ref VBuffer<uint> recoItems, ref VBuffer<float> recoScores) =>
+                {
+                    scores.Clear();
+                    if (!includeHistory)
+                    {
+                        history.Clear();
+                        for (int i = 0; i < items.Count; i++)
+                            history.Add(items.Values[i]);
+                    }
+
+                    int recoCount = Math.Min(_n, recommendationCount);
+                    if (userId == 0)
```

Comment:
>userId [](start = 24, length = 6)

any concerns of userId being <0?

Reply:
The entire function is removed.

---
In reply to: [225357420](https://github.com/dotnet/machinelearning/pull/1263#discussion_r225357420) [](ancestors = 225357420)